src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3888,61 +3905,70 @@ private void AddDisposingFinally(IOperation resource, bool knownToImplementIDisp
             EnterRegion(finallyRegion);
             AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block));
 
-            if (!knownToImplementIDisposable)
-            {
-                Debug.Assert(!isNotNullableValueType(resource.Type));
-                resource = ConvertToIDisposable(resource, iDisposable, isTryCast: true);
-                int captureId = GetNextCaptureId(finallyRegion);
-                AddStatement(new FlowCaptureOperation(captureId, resource.Syntax, resource));
-                resource = GetCaptureReference(captureId, resource);
-            }
-
-            if (!knownToImplementIDisposable || !isNotNullableValueType(resource.Type))
-            {
-                IOperation condition = MakeIsNullOperation(OperationCloner.CloneOperation(resource));
-                ConditionalBranch(condition, jumpIfTrue: true, endOfFinally);
-                _currentBasicBlock = null;
-            }
+            // In error recovery scenarios, we may not have a dispose method.
+            // We'll attempt to make a best guess at the interface, if it's available
+            disposeMethod ??= (isAsynchronous
+                                ? _compilation.CommonGetWellKnownTypeMember(WellKnownMember.System_IAsyncDisposable__DisposeAsync)
+                                : _compilation.CommonGetSpecialTypeMember(SpecialMember.System_IDisposable__Dispose)
+                              )?.GetISymbol() as IMethodSymbol;
 
-            if (!iDisposable.Equals(resource.Type))
+            if (disposeMethod is object)
```

Comment:
>if (disposeMethod is object) [](start = 12, length = 28)

It is quite possible that I am missing something obvious, but it feels to me that the code in this method was rewritten without actual necessity to do that. For example, addition of this ```if``` looks unnecessary, we handled the situation when there is no method symbol before this change.  It feels to me that all what was needed is just a small change to the tryDispose local function and to the ```if``` above its invocation, which is responsible for the conversion of the resource. I suggest to try starting from the original state of the function and making minimal changes only if necessary.